### PR TITLE
434 Add Command to Tail -n Lines

### LINKED ISSUE #434: Add command to tail n lines of output
**Is your feature request related to a problem? Please describe.**

I'd love a new gum command that allows a user to stream the last `n` lines of output — very much like how Docker displays steps in the build process.

**Describe the solution you'd like**

I think the following would be an intuitive interface for users to show the `12` most recent lines of output of `brew upgrade`:

```bash
brew upgrade | gum stream --lines 12 --faint --italic # etc …
```

**Describe alternatives you've considered**

I'm not really sure how else to do this on the command line. I suppose a curses library could do it but that's not super convenient

**Additional context**

N/A

### PR DESCRIPTION
**Pull request by CodeZea1ot**
Closes #434

### Changes
- Adds a new `gum tail` subcommand
- Allows user to stream stdin for -n number of lines
- Keeps logs from piped running processes from cluttering up the terminal

### Example
```bash
{ sudo apt update -y && sudo apt upgrade -y && sudo apt autoremove -y; } | gum tail -n 3 // Stream the last 3 lines of info generated by this process until complete
```

### CODE REVIEW

**Comment by maaslalani** [comment]
Hey @CodeZea1ot, what benefit do you get with `gum tail` vs simply using `tail`?

**Comment by CodeZea1ot** [comment]
Hey there @maaslalani 

The idea behind a `gum tail` is to be able to stream output from a long running process that produces a lot of output with a configurable amount of visible max lines.

For example, say you have a script that runs an install or download process that you don't want to hide with `gum spin`, you could instead show the last `-n` lines of the output to the user, with it refreshing until the process ends.

In other words, you would get up to `-n` lines of output at all times and those lines that are displayed would always show the latest lines from the piped command coming into stdin.


https://github.com/charmbracelet/gum/assets/54991101/f6030259-a3df-42ec-a750-2232bff30589

**Comment by newcron** [comment]
I have to say, that's one of the features I'm waiting for so that I can start to use it. 

My usecase is a local dev environment in which several automated steps are performed to simplify the dev process. 

in my case, I would love exactly that for my build script, which should: build the code, create a test environment (databases with fixtures,...), then start the application and then run the end to end tests. 

Each of these steps produces output, which I don't want the whole screen to pollute. at the same time, it would be very useful for the developers to get a glimpse on what's going on (e.g. if there is an error in the build, or the application crashes at startup). 

Filling the whole screen with that just feels too verbose, yet not showing at all leaves the user without any sense of real progress.

**Comment by CodeZea1ot** [comment]
I built a BASH CLI called `spyglass` for this use case. You can check it out over at https://github.com/CodeZea1ot/spyglass

This is the functionality I'd like to bring to a `gum tail` command.

main_demo

File: gum.go [hunk 1]
```diff
@@ -196,6 +197,13 @@ type Gum struct {
 	//
 	Table table.Options `cmd:"" help:"Render a table of data"`
 
+	// Tail
+	// Provides a means of streaming -n of lines from stdin
```
Comment (caarlos0) [inline gum.go:201]:
```suggestion
	// Tail provides a means of streaming a number of lines from stdin.
```

File: gum.go [hunk 2]
```diff
@@ -196,6 +197,13 @@ type Gum struct {
...
+	// Useful for showing a subset of output from a process without clogging up the terminal
```
Comment (caarlos0) [inline gum.go:202]:
```suggestion
	// It is useful for showing a subset of output from a process without clogging up the terminal.
```

File: gum.go [hunk 3]
```diff
@@ -196,6 +197,13 @@ type Gum struct {
...
+	//
+	// $ { sudo apt update -y && sudo apt upgrade -y && sudo apt autoremove -y; } | gum tail -n 3
```
Comment (caarlos0) [inline gum.go:204]:
```suggestion
	// Let's watch apt update:
	// 
	// $ { sudo apt update -y && sudo apt upgrade -y && sudo apt autoremove -y; } | gum tail -n 3
```

File: tail/command.go [hunk 4]
```diff
@@ -0,0 +1,62 @@
+package tail
+
+import (
+	"bufio"
+	"fmt"
+	"os"
+)
+
+// Run executes the tail command
```
Comment (caarlos0) [inline tail/command.go:9]:
```suggestion
// Run executes the tail command.
```

File: tail/command.go [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+func appendLine(lines []string, line string, maxLines int) []string {
+    lines = append(lines, line)
+    if len(lines) > maxLines {
+        lines = lines[1:]
+    }
+    return lines
+}
+
+func clearScreen() {
+    fmt.Print("\033[H\033[2J")
```
Comment (caarlos0) [inline tail/command.go:46]:
```suggestion
	fmt.Print(ansi.EraseDisplay(2))
```

File: tail/command.go [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+	fmt.Println()
+
+    // Print the last n lines
+    start := len(lines) - n
+    if start < 0 {
+        start = 0
+    }
+    for i := start; i < len(lines); i++ {
+        fmt.Println(lines[i])
+    }
```
Comment (caarlos0) [inline tail/command.go:61]:
```suggestion
    fmt.Println(strings.Join(lines, "\n"))
```

**Review by caarlos0 — COMMENTED** [review]
made some suggestions

also would be good to run goimports/gofmt

**Comment by reitzig** [comment]
Is there anything beyond above comments by @caarlos0  preventing this from being merged?

**Comment by KyleChamberlin** [comment]
I was just looking for this functionality in gum, I was sure I must just be missing the docs for this, but was surprised there wasn't something like this. Any chance this can get merged before the next release?
